runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp

```diff
@@ -366,111 +492,140 @@ getClassSignature(J9VMThread *currentThread, J9Class * clazz)
 		J9Class *pclass = J9VM_J9CLASS_FROM_HEAPCLASS(currentThread, pObject);
-		signatures[i] = getClassSignature(currentThread, pclass);
-		if (NULL == signatures[i]) {
-			goto done;
-		}
-
-		signatureLength += strlen(signatures[i]);
+		signatureLength += getClassSignatureLength(currentThread, pclass);
 	}
+	j9object_t rtype = J9VMJAVALANGINVOKEMETHODTYPE_RTYPE(currentThread, typeObject);
+	J9Class *rclass = J9VM_J9CLASS_FROM_HEAPCLASS(currentThread, rtype);
+	signatureLength = getClassSignatureLength(currentThread, rclass);
 
-	{
-		/* Return type */
-		j9object_t rtype = J9VMJAVALANGINVOKEMETHODTYPE_RTYPE(currentThread, typeObject);
-		J9Class *rclass = J9VM_J9CLASS_FROM_HEAPCLASS(currentThread, rtype);
-		rSignature = getClassSignature(currentThread, rclass);
-		if (NULL == rSignature) {
-			goto done;
+	result = reinterpret_cast<J9UTF8 *>(j9mem_allocate_memory(signatureLength + sizeof(J9UTF8), OMRMEM_CATEGORY_VM));
+	if (NULL != result) {
+		LocalJ9UTF8Buffer stringBuffer = LocalJ9UTF8Buffer(result, signatureLength);
+
+		stringBuffer.putCharAtCursor('(');
+		for (U_32 i = 0; i < numArgs; i++) {
+			j9object_t pObject = J9JAVAARRAYOFOBJECT_LOAD(currentThread, ptypes, i);
+			J9Class *pclass = J9VM_J9CLASS_FROM_HEAPCLASS(currentThread, pObject);
+			getClassSignatureInout(currentThread, pclass, &stringBuffer);
 		}
+		stringBuffer.putCharAtCursor(')');
 
-		signatureLength += strlen(rSignature);
+		j9object_t rtype = J9VMJAVALANGINVOKEMETHODTYPE_RTYPE(currentThread, typeObject);
+		J9Class *rclass = J9VM_J9CLASS_FROM_HEAPCLASS(currentThread, rtype);
+		getClassSignatureInout(currentThread, rclass, &stringBuffer);
+		stringBuffer.commitLength();
 	}
 
-	methodDescriptor = (J9UTF8*)j9mem_allocate_memory(signatureLength + sizeof(J9UTF8), OMRMEM_CATEGORY_VM);
-	if (NULL == methodDescriptor) {
-		goto done;
-	}
-	J9UTF8_SET_LENGTH(methodDescriptor, (U_16)signatureLength);
-	cursor = (char*)J9UTF8_DATA(methodDescriptor);
-	*cursor++ = '(';
+	return result;
+}
+
+/**
+ * Attempts to fill in a method signature constructed from a MethodType into a passed in signature buffer.
+ * Falls back to a dynamic buffer allocation if the statically allocated buffer's capacity is exceeded.
+ * @param[in] currentThread The J9VMThread of the current thread
+ * @param[in] typeObject A MethodType object that contains parameter and return type information
+ * @param[in,out] stringBuffer The signature buffer to place the signature into
+ * @return pointer to a filled-in JUTF8 signature buffer, either the dynamically or statically allocated buffer
+ */
+static J9UTF8 *
+getJ9UTF8SignatureFromMethodType(J9VMThread *currentThread, j9object_t typeObject, LocalJ9UTF8Buffer *stringBuffer)
+{
+	j9object_t ptypes = J9VMJAVALANGINVOKEMETHODTYPE_PTYPES(currentThread, typeObject);
+	U_32 numArgs = J9INDEXABLEOBJECT_SIZE(currentThread, ptypes);
+
+	/* Need space for at least '(', ')', '\0'. */
+	Assert_JCL_true(3 <= stringBuffer->capacity);
```

Comment:
not needed as `getJ9UTF8SignatureFromMethodType` always get called with 256 buffer size.